src/Psalm/Internal/Provider/ReturnTypeProvider/ArrayPointerAdjustmentReturnTypeProvider.php

```diff
@@ -44,15 +44,27 @@ public static function getFunctionReturnType(
             return Type::getMixed();
         }
 
+        $isNotEmpty = false;
+
         if ($first_arg_array instanceof Type\Atomic\TArray) {
             $value_type = clone $first_arg_array->type_params[1];
+            $isNotEmpty = $first_arg_array instanceof Type\Atomic\TNonEmptyArray;
         } elseif ($first_arg_array instanceof Type\Atomic\TList) {
             $value_type = clone $first_arg_array->type_param;
+            $isNotEmpty = $first_arg_array instanceof Type\Atomic\TNonEmptyList;
         } else {
             $value_type = $first_arg_array->getGenericValueType();
+            $isNotEmpty = $first_arg_array->getGenericArrayType() instanceof Type\Atomic\TNonEmptyArray;
+        }
+
+        if ($value_type->isEmpty()) {
+            $value_type->removeType('empty');
+        }
+
+        if (($function_id !== 'reset' && $function_id !== 'end') || !$isNotEmpty) {
+            $value_type->addType(new Type\Atomic\TFalse);
         }
```

Comment:
```suggestion
        if (($function_id !== 'reset' && $function_id !== 'end') || !$isNotEmpty) {
            $value_type->removeType('empty');
            $value_type->addType(new Type\Atomic\TFalse);
        }
```
This may be a better solution...